File: chapter02/main/src/main/java/game/Card.java
package game;

public class

Card {

  private static final int SPADE = 0;
  private static final int HEART = 1;
  private static final int DIAMOND = 2;
  private static final int CLUB = 3;
  private static final String[] suitEmoji = {"♠", "♥", "♦", "♣"};
  private static final String[] suit = {"S", "H", "D", "C"};
  private static final String[] resouseId = {"A", "2", "3", "4", "5", "6", "7", "8", "9", "T", "J",
      "Q", "K"};
  private static final String[] valueID = {"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J",
      "Q", "K"};
  private int type;
  private int value;

  public Card(int type, int value) {
    assert value >= 1 && value <= 13;
    this.type = type;
    this.value = value;
  }

  public String getSuit(boolean needEmoji) {
    return needEmoji ? suitEmoji[type] : suit[type];
  }

  public String getResourceName() {
    return getSuit(false) + resouseId[value - 1];
  }

  public void setType(int type) {
    this.type = type;
  }

  public int getValue() {
    return value;
  }

  public void setValue(int value) {
    this.value = value;
  }

  @Override
  public String toString() {
    return getSuit(true) + valueID[getValue() - 1];
  }
}

File: chapter02/main/src/main/java/game/Card.kt
package game

class Card(private var type: Int, var value: Int) {

    val resourceName: String
        get() = getSuit(false) + resouseId[value - 1]

    init {
        assert(value in 1..13)
    }

    fun getSuit(needEmoji: Boolean): String {
        return if (needEmoji) suitEmoji[type] else suit[type]
    }

    fun setType(type: Int) {
        this.type = type
    }

    override fun toString(): String {
        return getSuit(true) + valueID[value - 1]
    }

    companion object {
        private val SPADE = 0
        private val HEART = 1
        private val DIAMOND = 2
        private val CLUB = 3
        private val suitEmoji = arrayOf("♠", "♥", "♦", "♣")
        private val suit = arrayOf("S", "H", "D", "C")
        private val resouseId =
            arrayOf("A", "2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K")
        private val valueID =
            arrayOf("A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K")
    }
}
